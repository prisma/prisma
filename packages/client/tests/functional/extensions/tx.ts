import { faker } from '@faker-js/faker'
import { copycat } from '@snaplet/copycat'
import { expectTypeOf } from 'expect-type'

import testMatrix from './_matrix'
// @ts-ignore
import type { PrismaClient } from './node_modules/@prisma/client'

declare let prisma: PrismaClient

const email = faker.internet.email()

testMatrix.setupTestSuite(
  (_suiteConfig, _suiteMeta, clientMeta) => {
    beforeEach(async () => {
      await prisma.post.deleteMany()
      await prisma.user.deleteMany()
      const { id: userId } = await prisma.user.create({
        data: {
          email,
          firstName: 'John',
          lastName: 'Smith',
        },
      })

      await prisma.post.create({
        data: {
          user: { connect: { id: userId } },
        },
      })
    })

    testIf(clientMeta.runtime !== 'edge')('extended client in tx can rollback via normal call', async () => {
      const xprisma = prisma.$extends({
        result: {
          user: {
            fullName: {
              needs: { firstName: true, lastName: true },
              compute(user) {
                return `${user.firstName} ${user.lastName}`
              },
            },
          },
        },
      })

      const result = xprisma.$transaction([
        xprisma.user.create({
          data: {
            id: copycat.uuid(1).replaceAll('-', '').slice(-24),
            email: 'jane@smith.com',
            firstName: 'Jane',
            lastName: 'Smith',
          },
        }),
        xprisma.user.create({
          data: {
            id: copycat.uuid(2).replaceAll('-', '').slice(-24),
            email: 'jane@smith.com',
            firstName: 'Jane',
            lastName: 'Smith',
          },
        }),
      ])

      await expect(result).rejects.toMatchPrismaErrorSnapshot()

      const users = await prisma.user.findMany({ where: { email: 'jane@smith.com' } })

      expect(users).toHaveLength(0)
    })

    testIf(clientMeta.runtime !== 'edge')('extended client in tx works via normal call', async () => {
      const xprisma = prisma.$extends({
        result: {
          user: {
            fullName: {
              needs: { firstName: true, lastName: true },
              compute(user) {
                return `${user.firstName} ${user.lastName}`
              },
            },
          },
        },
      })

      await xprisma.$transaction([
        xprisma.user.create({
          data: {
            email: 'jane@smith.com',
            firstName: 'Jane',
            lastName: 'Smith',
          },
        }),
      ])

      const users = await prisma.user.findMany({ where: { email: 'jane@smith.com' } })

      expect(users).toHaveLength(1)
    })

    testIf(clientMeta.runtime !== 'edge')('extended client in tx can rollback via custom call', async () => {
      const xprisma = prisma
        .$extends({
          result: {
            user: {
              fullName: {
                needs: { firstName: true, lastName: true },
                compute(user) {
                  return `${user.firstName} ${user.lastName}`
                },
              },
            },
          },
        })
        .$extends({
          model: {
            $allModels: {
              createAlt(args: any) {
                return (this as any).create(args)
              },
            },
          },
        })

      const result = xprisma.$transaction([
        xprisma.user.createAlt({
          data: {
            id: copycat.uuid(1).replaceAll('-', '').slice(-24),
            email: 'jane@smith.com',
            firstName: 'Jane',
            lastName: 'Smith',
          },
        }),
        xprisma.user.createAlt({
          data: {
            id: copycat.uuid(2).replaceAll('-', '').slice(-24),
            email: 'jane@smith.com',
            firstName: 'Jane',
            lastName: 'Smith',
          },
        }),
      ])

      await expect(result).rejects.toMatchPrismaErrorSnapshot()

      const users = await prisma.user.findMany({ where: { email: 'jane@smith.com' } })

      expect(users).toHaveLength(0)
    })

    test('extended client in tx works via custom call', async () => {
      const xprisma = prisma
        .$extends({
          result: {
            user: {
              fullName: {
                needs: { firstName: true, lastName: true },
                compute(user) {
                  return `${user.firstName} ${user.lastName}`
                },
              },
            },
          },
        })
        .$extends({
          model: {
            $allModels: {
              createAlt(args: any) {
                return (this as any).create(args)
              },
            },
          },
        })

      await xprisma.$transaction([
        xprisma.user.create({
          data: {
            email: 'jane@smith.com',
            firstName: 'Jane',
            lastName: 'Smith',
          },
        }),
      ])

      const users = await prisma.user.findMany({ where: { email: 'jane@smith.com' } })

      expect(users).toHaveLength(1)
    })

    testIf(clientMeta.runtime !== 'edge')('isolation level is properly reflected in extended client', () => {
      ;async () => {
        const xprisma = prisma.$extends({})

        // @ts-test-if: provider !== Providers.MONGODB
        const data = await xprisma.$transaction([xprisma.user.findFirst({ select: { id: true } })], {
          isolationLevel: 'Serializable',
        })

        expectTypeOf(data).toEqualTypeOf<[{ id: string } | null]>()
      }
    })

    test('type inference allows for destructuring the array', () => {
      ;async () => {
        const xprisma = prisma.$extends({})

        const [data, count] = await xprisma.$transaction([
          xprisma.user.findFirst({ select: { id: true } }),
          xprisma.user.count(),
        ])

        expectTypeOf(data).toEqualTypeOf<{ id: string } | null>()
        expectTypeOf(count).toEqualTypeOf<number>()
      }
    })
  },
  {
    skipDriverAdapter: {
      from: ['js_d1'],
      reason:
        'batch transaction needs to be implemented. Unskip once https://github.com/prisma/team-orm/issues/997 is done',
    },
  },
)
