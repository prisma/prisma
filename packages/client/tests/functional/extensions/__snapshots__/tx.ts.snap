// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extensions.tx (provider=cockroachdb) extended client in tx can rollback via custom call 1`] = `
"
Invalid \`prisma.user.create()\` invocation:


Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=cockroachdb) extended client in tx can rollback via normal call 1`] = `
"
Invalid \`xprisma.user.create()\` invocation in
/client/tests/functional/extensions/tx.ts:0:0

  XX     lastName: 'Smith',
  XX   },
  XX }),
→ XX xprisma.user.create(
Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=mongodb) extended client in tx can rollback via custom call 1`] = `
"
Invalid \`prisma.user.create()\` invocation:


Unique constraint failed on the constraint: \`User_email_key\`"
`;

exports[`extensions.tx (provider=mongodb) extended client in tx can rollback via normal call 1`] = `
"
Invalid \`xprisma.user.create()\` invocation in
/client/tests/functional/extensions/tx.ts:0:0

  XX     lastName: 'Smith',
  XX   },
  XX }),
→ XX xprisma.user.create(
Unique constraint failed on the constraint: \`User_email_key\`"
`;

exports[`extensions.tx (provider=mysql) extended client in tx can rollback via custom call 1`] = `
"
Invalid \`prisma.user.create()\` invocation:


Unique constraint failed on the constraint: \`User_email_key\`"
`;

exports[`extensions.tx (provider=mysql) extended client in tx can rollback via normal call 1`] = `
"
Invalid \`xprisma.user.create()\` invocation in
/client/tests/functional/extensions/tx.ts:0:0

  XX     lastName: 'Smith',
  XX   },
  XX }),
→ XX xprisma.user.create(
Unique constraint failed on the constraint: \`User_email_key\`"
`;

exports[`extensions.tx (provider=mysql, js_planetscale) extended client in tx can rollback via custom call 1`] = `
"
Invalid \`prisma.user.create()\` invocation:


Unique constraint failed on the (not available)"
`;

exports[`extensions.tx (provider=mysql, js_planetscale) extended client in tx can rollback via normal call 1`] = `
"
Invalid \`xprisma.user.create()\` invocation in
/client/tests/functional/extensions/tx.ts:0:0

  XX     lastName: 'Smith',
  XX   },
  XX }),
→ XX xprisma.user.create(
Unique constraint failed on the (not available)"
`;

exports[`extensions.tx (provider=postgresql) extended client in tx can rollback via custom call 1`] = `
"
Invalid \`prisma.user.create()\` invocation:


Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=postgresql) extended client in tx can rollback via normal call 1`] = `
"
Invalid \`xprisma.user.create()\` invocation in
/client/tests/functional/extensions/tx.ts:0:0

  XX     lastName: 'Smith',
  XX   },
  XX }),
→ XX xprisma.user.create(
Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=postgresql, js_neon) extended client in tx can rollback via custom call 1`] = `
"
Invalid \`prisma.user.create()\` invocation:


Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=postgresql, js_neon) extended client in tx can rollback via normal call 1`] = `
"
Invalid \`xprisma.user.create()\` invocation in
/client/tests/functional/extensions/tx.ts:0:0

  XX     lastName: 'Smith',
  XX   },
  XX }),
→ XX xprisma.user.create(
Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=postgresql, js_pg) extended client in tx can rollback via custom call 1`] = `
"
Invalid \`prisma.user.create()\` invocation:


Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=postgresql, js_pg) extended client in tx can rollback via normal call 1`] = `
"
Invalid \`xprisma.user.create()\` invocation in
/client/tests/functional/extensions/tx.ts:0:0

  XX     lastName: 'Smith',
  XX   },
  XX }),
→ XX xprisma.user.create(
Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=sqlite) extended client in tx can rollback via custom call 1`] = `
"
Invalid \`prisma.user.create()\` invocation:


Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=sqlite) extended client in tx can rollback via normal call 1`] = `
"
Invalid \`xprisma.user.create()\` invocation in
/client/tests/functional/extensions/tx.ts:0:0

  XX     lastName: 'Smith',
  XX   },
  XX }),
→ XX xprisma.user.create(
Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=sqlite, js_libsql) extended client in tx can rollback via custom call 1`] = `
"
Invalid \`prisma.user.create()\` invocation:


Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=sqlite, js_libsql) extended client in tx can rollback via normal call 1`] = `
"
Invalid \`xprisma.user.create()\` invocation in
/client/tests/functional/extensions/tx.ts:0:0

  XX     lastName: 'Smith',
  XX   },
  XX }),
→ XX xprisma.user.create(
Unique constraint failed on the fields: (\`email\`)"
`;

exports[`extensions.tx (provider=sqlserver) extended client in tx can rollback via custom call 1`] = `
"
Invalid \`prisma.user.create()\` invocation:


Unique constraint failed on the constraint: \`dbo.User\`"
`;

exports[`extensions.tx (provider=sqlserver) extended client in tx can rollback via normal call 1`] = `
"
Invalid \`xprisma.user.create()\` invocation in
/client/tests/functional/extensions/tx.ts:0:0

  XX     lastName: 'Smith',
  XX   },
  XX }),
→ XX xprisma.user.create(
Unique constraint failed on the constraint: \`dbo.User\`"
`;
